src/jdk.hotspot.agent/share/classes/sun/jvm/hotspot/oops/InstanceKlass.java

```diff
@@ -264,81 +258,71 @@ public long getSize() { // in number of bytes
-      return getConstants().getSymbolAt(nameIndex);
+    // Cannot use getFieldNameIndex() because this method is also used for injected fields
+    return getField(index).getName();
+  }
+
+  public Symbol getSymbolFromIndex(int cpIndex, boolean injected) {
+    if (injected) {
+      return vmSymbols.symbolAt(cpIndex);
     } else {
-      return vmSymbols.symbolAt(nameIndex);
+      return getConstants().getSymbolAt(cpIndex);
     }
   }
 
   public short getFieldSignatureIndex(int index) {
     if (index >= getJavaFieldsCount()) throw new IndexOutOfBoundsException("not a Java field;");
-    return getFields().at(index * FIELD_SLOTS + SIGNATURE_INDEX_OFFSET);
+    return (short)getField(index).getGenericSignatureIndex();
   }
 
   public Symbol getFieldSignature(int index) {
-    int signatureIndex = getFields().at(index * FIELD_SLOTS + SIGNATURE_INDEX_OFFSET);
-    if (index < getJavaFieldsCount()) {
-      return getConstants().getSymbolAt(signatureIndex);
-    } else {
-      return vmSymbols.symbolAt(signatureIndex);
-    }
+    // Cannot use getFieldSignatureIndex() because this method is also use for injected fields
+    return getField(index).getSignature();
   }
 
   public short getFieldGenericSignatureIndex(int index) {
-    // int len = getFields().length();
-    int allFieldsCount = getAllFieldsCount();
-    int generic_signature_slot = allFieldsCount * FIELD_SLOTS;
-    for (int i = 0; i < allFieldsCount; i++) {
-      short flags = getFieldAccessFlags(i);
-      AccessFlags access = new AccessFlags(flags);
-      if (i == index) {
-        if (access.fieldHasGenericSignature()) {
-           return getFields().at(generic_signature_slot);
-        } else {
-          return 0;
-        }
-      } else {
-        if (access.fieldHasGenericSignature()) {
-          generic_signature_slot ++;
-        }
-      }
-    }
-    return 0;
+    return getField(index).getGenericSignatureIndex();
   }
 
   public Symbol getFieldGenericSignature(int index) {
-    short genericSignatureIndex = getFieldGenericSignatureIndex(index);
-    if (genericSignatureIndex != 0)  {
-      return getConstants().getSymbolAt(genericSignatureIndex);
-    }
-    return null;
+    return getField(index).getGenericSignature();
   }
 
   public short getFieldInitialValueIndex(int index) {
     if (index >= getJavaFieldsCount()) throw new IndexOutOfBoundsException("not a Java field;");
-    return getFields().at(index * FIELD_SLOTS + INITVAL_INDEX_OFFSET);
+    return (short)getField(index).getInitialValueIndex();
   }
 
   public int getFieldOffset(int index) {
-    U2Array fields = getFields();
-    short lo = fields.at(index * FIELD_SLOTS + LOW_OFFSET);
-    short hi = fields.at(index * FIELD_SLOTS + HIGH_OFFSET);
-    if ((lo & FIELDINFO_TAG_OFFSET) == FIELDINFO_TAG_OFFSET) {
-      return VM.getVM().buildIntFromShorts(lo, hi) >> FIELDINFO_TAG_SIZE;
-    }
-    throw new RuntimeException("should not reach here");
+    return (int)getField(index).getOffset();
```

Comment:
Cast to int is not needed

Reply:
Other APIs (like MetadaField) are using longs to pass offsets, doing a cast here is less disruptive than changing all the other APIs.